STAR.py

```diff
@@ -46,101 +77,140 @@ def fivestar(ballots: pd.DataFrame):
 ##############################################################################
 
-def default_scoring_tiebreaker(ballots: pd.DataFrame):
-    tmp = weak_condorcet_winners(ballots)
+def default_scoring_tiebreaker(pairwise_matrix: pd.DataFrame):
+    tmp = weak_condorcet_winners(pairwise_matrix)
     if len(tmp) == 1:
         return tmp[0]
     elif len(tmp) == 0:
-        return TrueTie(list(ballots.columns))
+        return TrueTie(list(pairwise_matrix.columns))
     else:
         return TrueTie(tmp)
 
-def default_runoff_tiebreaker(ballots: pd.DataFrame):
+def default_runoff_tiebreaker(summar_data: SummaryData):
     # Any pre-runoff ties must be resolved
-    assert len(ballots.columns) == 2
+    assert len(summar_data.score_sums.index) == 2
 
-    a,b = ballots.columns
-    scores = ballots.sum()
+    a,b = summar_data.score_sums.index
+    scores = summar_data.score_sums
     if np.isclose(scores[a], scores[b]):
-        return TrueTie(list(ballots.columns))
+        return TrueTie(list(summar_data.score_sums.index))
     else:
         return scores.argmax()
 
-def STAR(ballots: pd.DataFrame, scoring_tiebreaker=default_scoring_tiebreaker, runoff_tiebreaker=default_runoff_tiebreaker):
-    """
-        Given an election instance and optionally specified tiebreakers, return the STAR winner.
-
-        Parameters:
-            ballots (pd.DataFrame): contains a dataframe oriented such that each column name is a candidate
-                and each row is a voter's ballot. That is, the score voter A gives to candidate X is ballots[X][A]
-            
-            scoring_tiebreaker (function): Given an election instance, returns a winner according to the tiebreaker
-                or a TrueTie instance if it is not sufficiently resolute. Called during the scoring round of STAR.
-            
-            runoff_tiebreaker (function): Given an election instance, returns a winner according to the tiebreaker
-                or a TrueTie instance if it is not sufficiently resolute. Called during the runoff round of STAR.
-    """
-
-    if ballots.shape[1] < 1:
-        raise InvalidElection("Not enough candidates to fill desired number of seats")
-
+def Run_STAR_Round(summary_data: SummaryData, scoring_tiebreaker=default_scoring_tiebreaker, runoff_tiebreaker=default_runoff_tiebreaker):
     # If there is only one candidate, elect them
-    if ballots.shape[1] == 1:
-        return ballots.columns[0]
-
+    
+    round_results = {'winners':[], 'runner_up':[], 'logs':[]}
+    
+    if len(summary_data.score_sums.index) == 1:
+        round_results['winners'] = summary_data.score_sums.index
+        return summary_data.score_sums.index
+ 
     runoff_candidates = []
     while len(runoff_candidates) < 2:
 
-        eligible = ballots.drop(runoff_candidates, axis=1)
-        w = scoring_tiebreaker(ballots[score_winners(eligible)])
-        if isinstance(w, TrueTie):
+        eligible = copy.deepcopy(summary_data).drop(runoff_candidates)
```

Comment:
not sure if copy is necessary, strictly speaking, but probably doesn't hurt

Reply:
I think it is needed here because the drop and keep methods do alter the object rather than returning the altered object, but I can try changing that.

Reply:
oh right, you're overloading drop. I was thinking it was just DataFrame.drop method. sg then